linera-core/src/updater.rs

```diff
@@ -211,21 +211,31 @@ where
             .handle_optimized_certificate(&certificate, delivery)
             .await;
 
-        match &result {
-            Err(original_err @ NodeError::BlobsNotFound(blob_ids)) => {
-                self.remote_node
-                    .check_blobs_not_found(&certificate, blob_ids)?;
+        if let Err(NodeError::BlobsNotFound(blob_ids)) = result {
+            self.local_node
+                .check_missing_blob_ids(&certificate.required_blob_ids(), &blob_ids)?;
 
-                let blobs = self
-                    .local_node
-                    .find_missing_blobs(blob_ids.clone(), certificate.inner().chain_id())
-                    .await?
-                    .ok_or_else(|| original_err.clone())?;
-                self.remote_node
-                    .handle_certificate(certificate, blobs, delivery)
-                    .await
+            let blobs = self
+                .local_node
+                .find_missing_blobs(blob_ids.clone(), &certificate)
+                .await?;
+            let missing_blob_ids = blobs
+                .iter()
+                .zip(&blob_ids)
+                .filter_map(|(maybe_blob, blob_id)| maybe_blob.is_none().then_some(*blob_id))
+                .collect::<Vec<_>>();
+            if !missing_blob_ids.is_empty() {
+                return Err(NodeError::LocalError {
+                    error: LocalNodeError::BlobsNotFound(missing_blob_ids).to_string(),
+                });
```

Comment:
This one should be fine, right? The validator was missing some blobs, then we try to find them in the local node, and some are _also_ missing from the local node, so we return a local error 🤔